Utils/JSONSchema/infer.h

```diff
@@ -0,0 +1,542 @@
+#define CURRENT_UTILS_JSONSCHEMA_INFER_H
+
+#include "../../TypeSystem/struct.h"
+#include "../../TypeSystem/Schema/schema.h"
+#include "../../TypeSystem/Serialization/json.h"
+
+namespace current {
+namespace utils {
+
+struct InferSchemaException : Exception {
+  InferSchemaException() : Exception() {}
+  explicit InferSchemaException(const std::string& message) : Exception(message) {}
+};
+
+// The input string is not a valid JSON.
+struct InferSchemaParseJSONException : InferSchemaException {};
+
+// The input JSON can not be a `CURRENT_STRUCT`.
+struct InferSchemaInputException : InferSchemaException {
+  InferSchemaInputException() : InferSchemaException() {}
+  explicit InferSchemaInputException(const std::string& message) : InferSchemaException(message) {}
+};
+
+struct InferSchemaTopLevelEmptyArrayIsNotAllowed : InferSchemaInputException {};
+struct InferSchemaArrayOfNullsOrEmptyArraysIsNotAllowed : InferSchemaInputException {};
+struct InferSchemaUnsupportedTypeException : InferSchemaInputException {};
+struct InferSchemaInvalidCPPIdentifierException : InferSchemaInputException {
+  explicit InferSchemaInvalidCPPIdentifierException(const std::string& id)
+      : InferSchemaInputException("Invalid C++ identifier used as the key in JSON: '" + id + "'.") {}
+};
+
+struct InferSchemaIncompatibleTypesBase : InferSchemaInputException {
+  InferSchemaIncompatibleTypesBase() : InferSchemaInputException("Incompatible types.") {}
+  explicit InferSchemaIncompatibleTypesBase(const std::string& message) : InferSchemaInputException(message) {}
+};
+
+template <typename T_LHS, typename T_RHS>
+struct InferSchemaIncompatibleTypes : InferSchemaIncompatibleTypesBase {
+  InferSchemaIncompatibleTypes(const T_LHS& lhs, const T_RHS& rhs)
+      : InferSchemaIncompatibleTypesBase("Incompatible types: '" + lhs.HumanReadableType() + "' and '" +
+                                         rhs.HumanReadableType() + "'.") {}
+};
+
+struct InferSchemaArrayAndObjectAreIncompatible : InferSchemaIncompatibleTypesBase {};
+
+namespace impl {
+
+inline bool IsValidCPPIdentifier(const std::string& s) {
+  if (s.empty()) {
+    return false;
+  }
+  if (s == "NULL" || s == "null" || s == "true" || s == "false") {
+    // TODO(dkorolev): Test for other reserved words, such as "case" or "int".
+    return false;
+  }
+  if (!(s[0] == '_' || std::isalpha(s[0]))) {
+    return false;
+  }
+  for (size_t i = 1u; i < s.length(); ++i) {
+    if (!std::isalnum(s[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+inline std::string MaybeOptionalHumanReadableType(bool has_nulls, const std::string& type) {
+  return (has_nulls ? "optional " : "") + type;
+}
+
+// Inferred JSON types for fields.
+// Internally, the type is maintained along with the hitsogram of its values seen.
+CURRENT_STRUCT(String) {
+  CURRENT_FIELD(values, (std::map<std::string, size_t>));
+  CURRENT_FIELD(instances, size_t, 1);
+  CURRENT_FIELD(nulls, size_t, 0);
```

Comment:
Good catch, my bad. 4bc59f4

And added a TODO to add this check into `CURRENT_FIELD` definition.